### PR TITLE
Reset connecting flag when using DNS resolve

### LINKED ISSUE #200: TCP client does not retry following failed connection attempt using DNS resolve
When using async_resolve the TCP client does not retry following failed connection attempt and is stuck in that state until the master is reset.

Will raise PR shortly which fixes the issue.

### PR DESCRIPTION
**Pull request by sidhoda**
Fixes #200

### CODE REVIEW

File: cpp/libs/src/asiopal/TCPClient.cpp [hunk 1]
```diff
@@ -123,7 +123,11 @@ void TCPClient::HandleResolveResult(
 		// attempt a connection to each endpoint in the iterator until we connect
 		auto cb = [self = shared_from_this(), callback](const std::error_code & ec, asio::ip::tcp::resolver::iterator endpoints)
 		{
-			callback(self->executor, std::move(self->socket), ec);
+			self->connecting = false;
```
Comment (sidhoda) [inline cpp/libs/src/asiopal/TCPClient.cpp:126]: There's some duplication here with the code above this. Possibly could extract out into a common method.

**Comment by sidhoda** [comment]
Also, should checking/setting connecting flag be moved to after calling shared_from_this in BeginConnect?

```
bool TCPClient::BeginConnect(const connect_callback_t& callback)
{
	if (connecting || canceled) return false;
        ...
	auto self = this->shared_from_this();
	if (connecting || canceled) return false;
	this->connecting = true;
```

https://github.com/sidhoda/dnp3/blob/a3043acecce30a3ebbf5e10974c0a3d81d6fa94c/cpp/libs/src/asiopal/TCPClient.cpp#L62

**Comment by sidhoda** [comment]
@jadamcrain What's your thoughts on adding a "Connecting to: %s" log message in here?
https://github.com/automatak/dnp3/blob/2.0.x/cpp/libs/src/asiodnp3/TCPClientIOHandler.cpp#L70

It would have made this easier to spot the issue in this case, although it could potentially bloat logs if lots of connection attempts are happening.

**Comment by jadamcrain** [comment]
I think adding the log message is fine. We already have a log message for connection failures so it doesn't appreciably change the number of messages. Also, connection failures "back-off" so fewer and fewer on dead endpoints.

I will take a peek at fix and flag issue when time permits.

**Comment by sidhoda** [comment]
Okay, I'll add the log message and update the PR.

I have a feeling I've misunderstood the use of shared_from_this, was thinking calling that method is like entering a critical section. Ignore my comment above.

**Comment by jadamcrain** [comment]
Thanks Sid. shared_from_this() is just used to grab a shared_ptr<T> that points back to the object. This ensures that the object is not deleted while an asynchronous callback is pending. It isn't a critical section. The run-time cost is the same as  shared_ptr, e.g. atomic increment/decrement.

**Comment by sidhoda** [comment]
All makes sense now, thanks.

**Comment by sidhoda** [comment]
Thanks @jadamcrain, would we be a position to tag a new RC?

**Comment by jadamcrain** [comment]
Done. 2.2.0-RC5 building on appveyor and update in changelog.

BTW, there's some nice improvements to the master scheduler for multi-drop in another branch, that we can bring into 2.0.x once we start on 2.2.1.

**Comment by sidhoda** [comment]
Thanks! Happy to give the master multi-drop a test when it's ready, we use this in a few places.

**Pull request merged**